### PR TITLE
Feature: new HTTP server metrics with full status code

### PR DESCRIPTION
**Pull request by gmaz42**
## Which problem is this PR solving?

Mainly:

* Prometheus metrics to have the full HTTP status code

We currently use Jaeger metrics which only sport `5xx`, `2xx` and the like; having the full status code in metrics is necessary for tailoring alerts and making more specific Grafana panels.

This PR models the new metrics around the existing server-side gRPC metrics.

Secondarily:
* allowing to disable Jaeger/OpenTracing metrics (NOTE: this also means that the failures at parsing `JAEGER_*` environment variables now would happen when `Run()` is called instead of the constructor)
* do not cache only the last `Write` in the response writer middleware, but all writes
* do not cache written response if unnecessary for error logging purposes

## Short description of the changes

We currently have:
1. `{microservice_name}_http_requests_total`, `{microservice_name}_http_requests_latency` - these are provided by the Jaeger package and have labels `endpoint="GET-/hello/world"` and `status_code="2xx"`
2.  `component_grpc_handled_total` and  `component_grpc_handled_seconds` with labels `grpc_code="OK"`, `grpc_method="CreateMyEvent"`, `grpc_service="myservice.Service"`, `grpc_type="unary"`

With this PR we would have:
1. same as (1) before, but can be disabled by using `WithoutJaeger()` option
2. same as (2) before
3. `component_http_handled_total` and  `component_http_handled_seconds` with labels `status_code="200"`, `method="GET"`, `path="/hello/world"`, which follows the style of (2)

### CODE REVIEW

**Comment by mantzas** [comment]
@gmaz42 the CI seems to be broken.

**Comment by gmaz42** [comment]
@mantzas now should be good :crossed_fingers: locally tests passed

File: component/http/middleware.go [hunk 1]
```diff
@@ -133,17 +139,74 @@ func NewAuthMiddleware(auth auth.Authenticator) MiddlewareFunc {
...
 			logRequestResponse(corID, lw, r)
-			statusCodeErrorLogging(r.Context(), statusCodeLogger, lw.Status(), lw.payload, path)
+			if log.Enabled(log.ErrorLevel) && statusCodeLogger.shouldLog(lw.status) {
+				log.FromContext(r.Context()).Errorf("%s %d error: %v", path, lw.status, string(lw.responsePayload.Bytes()))
+			}
+		})
+	}
+}
+
+var (
```
Comment (gmaz42) [inline component/http/middleware.go]: I can move this block to the top of the file (after constants) if we prefer that style.

File: trace/trace.go [hunk 2]
```diff
@@ -48,7 +48,7 @@ func Setup(name, ver, agent, typ string, prm float64, buckets []float64) error {
 			LocalAgentHostPort:  agent,
 		},
 	}
-	time.Sleep(100 * time.Millisecond)
```
Comment (gmaz42) [inline trace/trace.go:51]: I thought this is unnecessary, but I can put it back since it is not strictly related to what I am implementing here.

**Comment by mantzas** [comment]
@gmaz42 This PR will introduce an additional `duplicate` metric to the service. Would it make sense to try to replace the existing integrated one with the new you are introducing? Trying here to keep things tight.

**Comment by gmaz42** [comment]
@mantzas the existing one is a side-effect of using Jaeger spans; I did not replace it because I thought some user might want to have Jaeger spans and perhaps they don't care about the metrics, but if we remove the spans the functionality will be lost. I am neutral on whether to remove it or not, but to be conservative here I added a method to disable the Jaeger integration while by default it is enabled.

Unless I am missing something and there is actually another metric somewhere else not implemented using Jaeger spans?

**Comment by gmaz42** [comment]
The CI is failing with this error:
```
API error (500): Head "https://registry-1.docker.io/v2/library/nats/manifests/2.1.7-scratch": received unexpected HTTP status: 503 Service Unavailable
```

But this is not due to any problem in this PR, rather due to the dependency on the Docker registry.

File: component/http/middleware.go (hunk 1 above)
Comment (mantzas) [reply component/http/middleware.go]: yes, this makes more sense.

File: trace/trace.go (hunk 2 above)
Comment (mantzas) [reply trace/trace.go:51]: It was added for a reason i assume which I do not remember TBH.

**Comment by mantzas** [comment]
> @mantzas the existing one is a side-effect of using Jaeger spans; I did not replace it because I thought some user might want to have Jaeger spans and perhaps they don't care about the metrics, but if we remove the spans the functionality will be lost. I am neutral on whether to remove it or not, but to be conservative here I added a method to disable the Jaeger integration while by default it is enabled.
> 
> Unless I am missing something and there is actually another metric somewhere else not implemented using Jaeger spans?

The idea is to have one of them and not two which is misleading and an unnecessary duplication.
Maybe we can produce the traces and the metrics manually and completely remove the existing integration.
We have though to be sure that the produced metrics are the same as the existing one because it will also lead to a breaking change. With this we will have the granularity you need and not support duplicate and confusing metrics. WDYT?

**Comment by gmaz42** [comment]
> We have though to be sure that the produced metrics are the same as the existing one because it will also lead to a breaking change. With this we will have the granularity you need and not support duplicate and confusing metrics. WDYT?

So instead of this metric which is similar to the gRPC one you would have only 1 that basically simulates how Jaeger currently does it, but with the extra full status code? Re-implemented not using the Jaeger package/integration. Is this what you have in mind?

**Comment by mantzas** [comment]
> > We have though to be sure that the produced metrics are the same as the existing one because it will also lead to a breaking change. With this we will have the granularity you need and not support duplicate and confusing metrics. WDYT?
> 
> So instead of this metric which is similar to the gRPC one you would have only 1 that basically simulates how Jaeger currently does it, but with the extra full status code? Re-implemented not using the Jaeger package/integration. Is this what you have in mind?

Yes, the end result would be one metric, ideally backwards compatible by name, and no need to add all the logic to use this or that metric. should be more consistent and not just added to suite one specific purpose. WDYT?

**Comment by gmaz42** [comment]
> Yes, the end result would be one metric, ideally backwards compatible by name, and no need to add all the logic to use this or that metric. should be more consistent and not just added to suite one specific purpose. WDYT?

@mantzas it is tricky because if it is backward compatible then it is not consistent with the gRPC one; let me make a recap: we currently have:
1. `{microservice_name}_http_requests_total`, `{microservice_name}_http_requests_latency` - these are provided by the Jaeger package and have labels `endpoint="GET-/hello/world"` and `status_code="2xx"`
2.  `component_grpc_handled_total` and  `component_grpc_handled_seconds` with labels `grpc_code="OK"`, `grpc_method="CreateMyEvent"`, `grpc_service="myservice.Service"`, `grpc_type="unary"`

These two are already inconsistent; with this PR we would have:
1. same as (1) before, but can be disabled by using `WithoutJaeger()` option
2. same as (2) before
3. `component_http_handled_total` and  `component_http_handled_seconds` with labels `status_code="200"`, `method="GET"`, `path="/hello/world"`

(3) would be consistent with (2); if we now make (3) look like (1) instead, should in future also adjust the gRPC one? :thinking: 

I have copied this recap in the PR description.

**Comment by gmaz42** [comment]
After talking about this we have opted for:
* introduce the new HTTP metric, consistent with how the gRPC one exists
* document how the Jaeger-provided HTTP metric and the new HTTP metric work
* do not allow disabling the Jaeger integration

The PR should now reflect these changes; I have also updated documentation accordingly.

**Review by mantzas — DISMISSED** [review]

**Review by teivah — APPROVED** [review]

**Review by mantzas — APPROVED** [review]

**Pull request merged**